Support remote IEx with NervesHub

jjcarstens opened the pull request:
Counterpart for https://github.com/nerves-hub/nerves_hub_web/pull/431

Creates a `ConsoleChannel` that can spawn and monitor an IEx Server.
It then ports that servers I/O back and forth over the Phoenix channel
to allow running a remote IEx session from a browser.

Leaving as a draft for a bit so others can experiment and help add/change breakpoints. Plus, will be adding tests soon

ConnorRigby commented:
I think the dialyzer issue here can be solved by adding 
```elixir
   def package() do
      [
        #...
        dialyzer: [plt_add_apps: [:iex]]
        #...
      ]
   end
```
to the mix.exs

fhunleth commented:
Instead of adding the app to the dialyxir options, I'd added it to the `extra_applications` list so that Distillery knows that it's a dependency as well. Not that Distillery wouldn't include ':iex', but since it actually is a dependency now, it feels better to say that it is.

ConnorRigby commented:
Does dialyzer know about it if you do that? If so i like that router better.

jjcarstens commented:
The issue was 2 part. To use `IEx.Server.run/1` in this way, we need Elixir 1.8. So I updated CI and mix.exs to require that in https://github.com/nerves-hub/nerves_hub/pull/86/commits/ec118c036c88a20256930a87d4a80adcfd653a12. The other is what was mentioned so I've also added `:iex` to the `extra_applications`

Also, finished the tests and updated doc for it. This should be ready for review

mobileoverlord commented:
This looks good, but, my only concern is that there is no way to opt out other than reconstructing the whole supervisor. Does anyone else feel like we should address that now?

ConnorRigby commented:
I briefly thought about this and decided not to worry about it. We could add it to application env maybe?

fhunleth commented:
Now that you bring this up, I'm for making this opt-in. This feature requires more trust in https://nerves-hub.org than other features. For example, firmware updates are signed, so it's not like we can substitute in malicious firmware. However, if we were evil, shell access would let us do anything, so it doesn't feel like a good default.

jjcarstens commented:
That is a good point and I like the idea of it being off by default. I’m thinking we have it establish the channel connection and use the Application env to decide if it’s allowed so that a remote IEx request from NervesHub can get some feedback that the device has it disabled (instead of trying to connect and the UI just not responding)

jjcarstens commented:
Updated. This now joins the `console` topic by default. But to start an IEx server and send messages back and forth, you need to have `config :nerves_hub, remote_iex: true`.
